### ISSUE TITLE
Client: FormData._gen_form_data ignores filename, explodes

### ISSUE THREAD

**Issue by patrick-jones** [aio-libs/aiohttp#2201]
I am using the client to upload a file using `FormData`, almost exactly the same code as shown in [this example](http://aiohttp.readthedocs.io/en/stable/client.html#post-a-multipart-encoded-file). The main difference is that I am passing a file created by [tempfile.TemporaryFile](https://docs.python.org/3/library/tempfile.html) as the second argument to `add_field`.

The result is: `TypeError: Can not serialize value type: <class '_io.BufferedRandom'>
 headers: {'Content-Type': 'application/vnd.ms-excel'}
 value: <_io.BufferedRandom name=23>`

The root exception is actually thrown from [`helpers.guess_filename`](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/helpers.py#L280): `TypeError: 'int' object is not subscriptable`

## Expected behaviour

There should be no need to call `guess_filename` because I specified a filename:
`form.add_field('file', data, filename=filename, content_type=content_type)`

And even if I didn't provide a filename, `guess_filename` should not throw an exception if `name` is not a string (which is apparently the case for tempfiles; notice that my tempfile has the number 23 for a name).

## Actual behaviour

[`FormData._gen_form_data`](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/formdata.py#L115-L117) is not passing the filename to `payload.get_payload`, even though the filename is available. It seems like the code could be changed to include `filename=dispparams.get('filename')` in the kwargs for `get_payload`.

`guess_filename` assumes that `name` will be a string. Oddly enough, sometimes it's a number.

## Steps to reproduce

Here is my code that produces the error. `data` is a `tempfile.TemporaryFile()`. The exception is thrown from the `post` call. `url`, `filename`, and `content_type` are all valid strings.

```python
    async def post_temp_file(self, url, filename, content_type, data):
        form = FormData()
        form.add_field('file', data, filename=filename, content_type=content_type)

        async with self.rest.post(url, data=form) as response:
            await self._raise_for_status(response)
            body = await response.text()

        return json.loads(body)
```

### Workaround

```python
        f = BytesIO(data.read())
        setattr(f, 'filename', filename)
        form.add_field('file', f, filename=filename, content_type=content_type)
```

## Your environment

Python 3.6.1 on Ubuntu

**Comment by patrick-jones** [comment]
Upon further inspection, it looks like it doesn't matter that `FormData._gen_form_data` doesn't pass the filename to `get_payload`. The filename is saved with a call to [`part.set_content_disposition`](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/formdata.py#L128) a few lines later. As long as `guess_filename` doesn't blow up everything is fine.

I will PR.

**Comment by asvetlov** [comment]
Fixed by #2207
